File: cpp/linked_list/linked_list.cpp
#include <iostream>

class Node {
public:
  int val_;
  Node *next = nullptr;

  Node(int val) { val_ = val; }
};

class LinkedList {

public:
  Node *head;
  Node *tail;

  LinkedList() {
    // initialize the list with a 'dummy' node which makes removing
    // a node from the beginning of a list easier.
    head = new Node(-1);
    tail = head;
  }

  void insertEnd(int val) {
    tail->next = new Node(val);
    tail = tail->next;
  }

  void remove(int index) {
    int i = 0;

    Node *current = head;

    while (i < index && current) {
      i++;
      current = current->next;
    }

    if (current && current->next) {
      if (current->next == tail) {
        tail = current;
      }

      current->next = current->next->next;
    }
  }

  void print() {
    Node *current = head->next;

    while (current) {
      std::cout << current->val_ << " -> ";
      current = current->next;
    }

    std::cout << std::endl;
  }
};

File: go/linkedlist/linkedlist.go
package linkedlist

import "fmt"

type Node struct {
	Val  int
	Next *Node
}

func NewNode(val int) *Node {
	return &Node{
		Val:  val,
		Next: nil,
	}
}

type LinkedList struct {
	Head *Node
	Tail *Node
}

func NewLinkedList() *LinkedList {
	node := NewNode(-1)

	return &LinkedList{
		Head: node,
		Tail: node,
	}
}

func (l *LinkedList) InsertEnd(val int) {
	l.Tail.Next = NewNode(val)
	l.Tail = l.Tail.Next
}

func (l *LinkedList) Remove(index int) {
	i := 0

	current := l.Head

	for i < index && current != nil {
		i++
		current = current.Next
	}

	if current != nil && current.Next != nil {
		if current.Next == l.Tail {
			l.Tail = current
		}

		current.Next = current.Next.Next
	}
}

func (l *LinkedList) Print() {
	current := l.Head.Next

	for current != nil {
		fmt.Printf("%d -> ", current.Val)
		current = current.Next
	}

	fmt.Println()
}
